### PR TITLE
[mixin] Moved download `crl-view` to mixin #116

### LINKED ISSUE #116: [mixins] Convert download `crl_view` to mixin
1. Move the download `crl_view` to mixin to allow reusability for REST API of `PKI` app in `openwisp-controller` module.
2. Presenty un-authenticated users are also able to download `CRL` of any CA.

https://github.com/openwisp/django-x509/blob/43d810f57ea592ece67a1b54904104cc022ca514/django_x509/base/admin.py#L135-L143

### PR DESCRIPTION
**Pull request by ManishShah120**
Fixes #116

### CODE REVIEW

File: django_x509/tests/test_ca.py [hunk 1]
```diff
@@ -13,13 +14,20 @@
...
 class TestCa(TestX509Mixin, TestCase):
     """
     tests for Ca model
     """
 
+    def setUp(self):
+        User.objects.create_superuser(
+            username='admin', password='tester', email='admin@admin.com'
+        )
+        self.client.login(username='admin', password='tester')
```
Comment (nemesisdesign) [inline django_x509/tests/test_ca.py:29]: why is this needed here?

**Review by nemesisdesign — COMMENTED** [review]
Looks good but I have a doubt, see below.

File: django_x509/tests/test_ca.py (hunk 1 above)
Comment (ManishShah120) [reply django_x509/tests/test_ca.py:29]: Hi, Previously It was implemented as:- only authenticated user should be able to download the CRL, But there was minor bug in the code, which allowed non authenticated user to download the CRL, and the tests written for it could not catch it. So after fixing it, for authenticating the user, we are required to login a User.

Comment (atb00ker) [reply django_x509/tests/test_ca.py:29]: Hmmm... isn't it required to be able to download CRL without auth, I mean for the network devices!?
Can you please confirm with existing behavior?

File: django_x509/base/admin.py [hunk 2]
```diff
@@ -132,16 +131,6 @@ def get_urls(self):
             url(r'^x509/ca/(?P<pk>[^/]+).crl$', self.crl_view, name='crl')
         ] + super().get_urls()
 
-    def crl_view(self, request, pk):
-        authenticated = request.user.is_authenticated
-        authenticated = authenticated() if callable(authenticated) else authenticated
-        if app_settings.CRL_PROTECTED and not authenticated:
```
Comment (ManishShah120) [inline django_x509/base/admin.py:138]: @atb00ker See here, it has been checked that the user is authenticated or not.,but in `if` condition shouldn't any one of them be enough for this condition to be true and forbid the user fro accessing it. I checked with this default code by coping the url and fetching it after logging out the user. It still allows to download.

File: django_x509/tests/test_ca.py (hunk 1 above)
Comment (ManishShah120) [reply django_x509/tests/test_ca.py:29]: https://github.com/openwisp/django-x509/pull/117/files#r643225106

File: django_x509/base/mixin.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    def crl_view(self, request, pk):
+        authenticated = request.user.is_authenticated
+        authenticated = authenticated() if callable(authenticated) else authenticated
+        if app_settings.CRL_PROTECTED or not authenticated:
+            return HttpResponse(_('Forbidden'), status=403, content_type='text/plain')
+        instance = get_object_or_404(self.model, pk=pk)
+        response = HttpResponse(
+            instance.crl, status=200, content_type='application/x-pem-file'
+        )
+        response['Content-Disposition'] = f'attachment; filename={pk}.crl'
```
Comment (nemesisdesign) [inline django_x509/base/mixin.py:18]: Why did you add this? In the previous code we didn't force the download.

File: django_x509/base/mixin.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+from django.utils.translation import ugettext_lazy as _
+
+from django_x509 import settings as app_settings
+
+
+class CrlDownloadMixin:
+    def crl_view(self, request, pk):
+        authenticated = request.user.is_authenticated
+        authenticated = authenticated() if callable(authenticated) else authenticated
+        if app_settings.CRL_PROTECTED or not authenticated:
```
Comment (nemesisdesign) [inline django_x509/base/mixin.py:12]: this is wrong, it should be like it was before.
By default there's no authentication, if `CRL_PROTECTED` is set to `True` then we shall check the user is authenticated.

File: django_x509/tests/test_ca.py (hunk 1 above)
Comment (nemesisdesign) [reply django_x509/tests/test_ca.py:29]: this test is not necessary, see my comments in the last reviews.
@atb00ker CRL is needed by software like OpenVPN, see `--crl-verify`in the openvpn manual: https://openvpn.net/community-resources/reference-manual-for-openvpn-2-4/.

**Review by nemesisdesign — CHANGES_REQUESTED** [review]

File: django_x509/base/mixin.py (hunk 4 above)
Comment (ManishShah120) [reply django_x509/base/mixin.py:12]: Ohh! Okay will undo it. :+1:

File: django_x509/base/mixin.py (hunk 3 above)
Comment (ManishShah120) [reply django_x509/base/mixin.py:18]: When this function is triggered by the API endpoint, then the file name doesn't get's added automatically instead the file in `crl`format get's downloaded without any name, Will try to implement it once again and see if I can do anything from the API part.

File: django_x509/base/admin.py (hunk 2 above)
Comment (nemesisdesign) [reply django_x509/base/admin.py:138]: https://github.com/openwisp/django-x509#django-x509-crl-protected
setting DJANGO_X509_CRL_PROTECTED to True will require authentication.

File: django_x509/base/mixin.py (hunk 3 above)
Comment (nemesisdesign) [reply django_x509/base/mixin.py]: ok, let's avoid changing the functionality here for reasons of backward compatibility.

**Review by nemesisdesign — APPROVED** [review]
